ENH: Added replay method to parse actions into 300s etc.

blimmo opened the pull request:
The new replay attribute `hits` contains a dictionary with lists of objects corresponding to 300s, 100s, 50s, misses and sliderbreaks hit calculated by comparing the actions to the replay's beatmap.

I've got it pretty accurate however occasionally things don't work currently. Mostly slider ticks are missed when they shouldn't be. I'm putting this here in the hope someone else will see what I'm missing while I work on it.

This output is all I want for my current project but I'm thinking others might want the replay action that made the hit (or was closest for a miss) to determine things like by how much they missed.  
I can't think of a good place to put that: the output structure might be complicated enough to deserve its own object? or should this data be in a separate method?).

llllllllll commented:
I haven't looked through this yet, but my use case for this information would be to build a time series of the user's accuracy though the beatmap. I believe this format will work for that. I will try to get through this soon.

blimmo commented:
I think the biggest reason some slider ticks are missed when they shouldn't be or vice versa is that I don't know how big the radius is to move your cursor when holding the slider. By my tests it's bigger than the visual cue, however it also seems to be a lot bigger if you're already holding the slider successfully.

llllllllll commented:
I would believe that it is larger than visually indicated. Anecdotally, I have had cases where it looked like I should slider break because my cursor was outside the circle but did not. I wonder if it is CS * some coefficient.

llllllllll commented on slider/beatmap.py:
```diff
@@ -1196,7 +1196,7 @@ def __init__(self,
         """The name of the map as it appears in game.
         """
-        return f'{self.artist} - {self.title} [{self.version}]]'
+        return f'{self.artist} - {self.title} [{self.version}]'
```
good catch!

llllllllll commented on slider/replay.py:
```diff
@@ -152,6 +154,64 @@ def _consume_actions(buffer):
+    return t.total_seconds() * 1e3
+
+
+def _within(p1, p2, d):
```
can you add a numpy-style docstring (the format used throughout the project) here explaining what `p1`, `p2`, and `d` are?

llllllllll commented on slider/replay.py:
```diff
@@ -576,6 +636,62 @@ def parse(cls, data, *, library=None, client=None, save=False):
+    def hits(self):
+        b = self.beatmap
+        rdata = self.actions
+        scores = {"perfects": [],
```
should we call these `'300', '100', '50'`?

blimmo replied:
forgot about those changes, will do

llllllllll commented on slider/replay.py:
```diff
@@ -576,6 +636,62 @@ def parse(cls, data, *, library=None, client=None, save=False):
+                  "goods": [],
+                  "bads": [],
+                  "misses": [],
+                  "sliderbreaks": [],
```
is `sliderbreak` normally one word? I am not actually sure, but if it isn't, could we split it like: `slider_breaks`?

blimmo replied:
i'm honestly not sure at all and a quick google gives both. i think 2 words is probably clearer but i really don't know. are they referenced anywhere else in the library already?

llllllllll replied:
This is the first time it is being used in the library; I don't have a strong preference so I will leave it to you.

llllllllll commented on slider/replay.py:
```diff
@@ -152,6 +154,64 @@ def _consume_actions(buffer):
+
+def _process_circle(obj, rdatum, hw, scores):
+    out_by = abs(_ms(rdatum.offset - obj.time))
+    if out_by < hw.hit_300:
```
should we just make the hit windows timedeltas so we can compare them to the offset directly?

blimmo replied:
yeah i forgot you could do that. I'm kinda reluctant to use timedeltas for things like this because i'm never sure how negatives and the like are handled and i am sure how ints work.

llllllllll replied:
timedeltas work pretty much like integers except they have as unit associated with them. You can have `- 5 seconds` or `+ 10 minutes` and then work with them together. It just makes sure that you don't accidentally mix seconds and minutes or something.

llllllllll commented on slider/replay.py:
```diff
@@ -152,6 +154,64 @@ def _consume_actions(buffer):
+        scores["perfects"].append(obj)
+    elif out_by < hw.hit_100:
+        scores["goods"].append(obj)
+    elif out_by < hw.hit_50:
```
what does the implicit, empty else case imply here? Is it not a miss?

blimmo replied:
It (should be) impossible to reach since this function is only reached when a hit is detected and hits are only looked for within the 50 hit window. So yes it is a miss but it's covered earlier on in the main function

llllllllll replied:
if that is the case, we should either remove the check here or add a final: `raise AssertionError('invalid circle')` so that it is clear that we cannot get there.

blimmo replied:
i like the `raise` option because i can see that saving me a lot of effort debugging but it's not that it's an invalid circle: that check should have already been done kind of indirectly so checking again is kinda misleading. I think i'm going to remove the last check

llllllllll commented on slider/replay.py:
```diff
@@ -152,6 +154,64 @@ def _consume_actions(buffer):
+    missed_points = 0
+    if not head_hit:
+        missed_points += 1
+        scores["sliderbreaks"].append(obj)
```
should this also go in the 300 section? Maybe we should return a `ReplayData` object (name tbd) which has properties that merge these or not.

blimmo replied:
hmm i'm not sure about an object, i was treating this more as something that sorts the hit objects in the map into categories and returning new objects instead feels a bit weird. Doesn't it also go into 300 or 100 or 50 or miss later? Although since it now must have non-zero `missed_points` it can't go into 300s anymore.

llllllllll commented on slider/replay.py:
```diff
@@ -576,6 +636,62 @@ def parse(cls, data, *, library=None, client=None, save=False):
 
+    @lazyval
+    def hits(self):
+        b = self.beatmap
```
please just call this `beatmap`; one letter variable names trip me up.

blimmo replied:
ah sorry that was left over from when i wasn't planning on including this in the library, sure thing.

llllllllll commented on slider/replay.py:
```diff
@@ -576,6 +636,62 @@ def parse(cls, data, *, library=None, client=None, save=False):
+    @lazyval
+    def hits(self):
+        b = self.beatmap
+        rdata = self.actions
```
it is confusing to give this a new name when binding locally, could we just call this `actions`, or rename both here and on the instance to `replay_data`?

blimmo replied:
yeah this is left over from before i put this code into the library (i really should check these things).

i was using replay_data because i think that's generally what it's called in the official osu info ([here](https://github.com/ppy/osu-api/wiki#get-replay-data)) but then again i don't think that name really makes much sense, i'd rather stick with actions

llllllllll commented on slider/replay.py:
```diff
@@ -576,6 +636,62 @@ def parse(cls, data, *, library=None, client=None, save=False):
+        rad = circle_radius(b.cs(easy=self.easy, hard_rock=self.hard_rock))
+        i = 0
+        for obj in b.hit_objects:
+            if self.hard_rock:
```
could we pull this body into a (maybe static) method which processes a single object? It makes the nesting of the loops easier to follow

blimmo replied:
hmm i hadn't really thought of doing that. Should i make the `process_slider` and `process_circle` functions static methods as well?

actually might that not cause problems with incrementing `i` (my pointer for the current action)? At the moment it's just incremented whenever i need to but in a separate function that could be tricky

llllllllll replied:
I guess it could return the pointer which gets forwarded to the next call. idk if that is cleaner. As far as moving things to statics they would be "private" like `_process_slider` but again, that may not be better. What do you think?

blimmo replied:
I do think that moving the stuff in the for loop into a separate function would probably be cleaner but i can't see a way to make the pointer work cleanly and it would also get passed a lot of things (like hw, rad etc.) that make it's signature really ugly. I was kinda thinking of the reasoning behind making a separate function being vaguely that someone might be able to use it on it's own.

I think that with a bit of a restructure it could work but I don't have time at the moment (and won't for a while). Is there a todo list i can put it on or something?

llllllllll replied:
idk if I would worry about it. If we can't remember to do this later then it isn't that important. I think what you have now works well so there is no need to refactor if it isn't natural.

blimmo commented:
I've addressed some of your comments. I wouldn't merge it yet since it doesn't work very well. I just realised the actual main `hits` lazyval doesn't have a docstring so if you do merge it you should probably mention it being unreliable there.

blimmo commented:
I've improved `hits` to the point where my none of my tests have more than 6 notes wrong. I'm still guessing numbers (you also seem to be able to take your cursor off the slider just before the end without missing the end tick and I don't know how far "much" is) and haven't had much luck with finding them in the source (if anyone knows where the slider logic is please share!).

I've changed the logic to check how far the cursor is at every replay action instead of only at the slider ticks. I'm also kind of guessing how this works so again the source would be really nice.

llllllllll commented on slider/replay.py:
```diff
@@ -152,6 +154,92 @@ def _consume_actions(buffer):
+    bool
+        Whether the distance between the points is less than d
+    """
+    return (p1.x - p2.x) ** 2 + (p1.y - p2.y) ** 2 < d ** 2
```
np.square is considerably faster than `**2`, idk if it matters here but may be good for future reference.

llllllllll commented on slider/replay.py:
```diff
@@ -152,6 +154,92 @@ def _consume_actions(buffer):
+                t_changes_append(t)
+                on = True
+
+    tick_ts = list(orange(obj.tick_rate, obj.num_beats, obj.tick_rate))
```
this function name cracks me up ;)

llllllllll commented:
Sorry about the delay. I also haven't been able to find this slider window stuff in the source after an hour or so of digging. Regarding your tests, please hold on to them, I plan on writing up a unit test suite soon and want some good use cases. We could even freeze a few replays into the source tree in `example_data.replays`, like how we have `example_data.beatmaps.sendan_life` which was just a beatmap I liked and use for a lot of testing.

I am pretty happy with this feature as is, we can always improve it later when we learn more about what the thresholds are from the source. One small nit pick is that `FIX:` commits should be `BUG:` or `ENH:`, but this could go in as is. Merge away!

blimmo commented:
I'm surprised you're ok with all the rubbish in my code but if it's ok with you I'll merge it! I've added a docstring and a few comments. I'm also starting to agree with you about making the hitwindow timedeltas but that belongs in another PR.

My test replays aren't really chosen to be representative (though I could probably rustle up some fairly good ones if I tried) I literally just copied out the first few replays from my replays folder.

llllllllll commented:
I think it is important to deliver incremental improvements. If adding this helps your use case then I am 100% okay with it. There will always be room for improvement, but it is best to get features in users' hands early so we can learn how to improve them later.

Thank you again for adding this extremely useful feature!
